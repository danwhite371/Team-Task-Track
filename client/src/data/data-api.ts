import type { OperationResult, Task } from '../types';
import {
  createTask,
  fetchTasks,
  startTask as startTaskRequest,
  stopTask as stopTaskRequest,
  fetchTaskTimes as fetchTaskTimesRequest,
} from './graphql';

export default class DataApi {
  updateTaskData: (tasks: Task[]) => void;
  updateOperationResult: (operationResult: OperationResult) => void;
  tasks: Task[] = [];

  constructor(
    updateTaskData: (tasks: Task[]) => void,
    updateOperationResult: (operationResult: OperationResult) => void
  ) {
    this.updateTaskData = updateTaskData;
    this.updateOperationResult = updateOperationResult;
    this.sendTasks();
  }

  // We want to send a message and a state, loading, success, failure
  async createNewTask(name: string) {
    this.updateOperationResult({ status: 'loading', message: 'Connecting...' });
    let task: Task;
    try {
      task = await createTask(name);
      console.log('createNewTask', JSON.stringify(task, null, 2));
      this.tasks.unshift(task);
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task created.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async sendTasks() {
    const tasks = await fetchTasks();
    this.tasks = tasks;
    this.updateTaskData(tasks);
  }

  // Get a new Task item with, replace current item with it, sort items by last time
  async startTask(id: number) {
    console.log('[DataApi] startTask', id);
    const task = await startTaskRequest(id);
    this.replaceTask(id, task);
    this.sortTasks();
    this.tasks = [...this.tasks];
    this.updateTaskData(this.tasks);
  }

  async stopTask(id: number) {
    console.log('[DataApi] stopTask', id);
    const task = await stopTaskRequest(id);
    console.log('[DataApi] stopTask', JSON.stringify(task, null, 2));
    this.replaceTask(id, task);
    this.sortTasks();
    this.tasks = [...this.tasks];
    this.updateTaskData(this.tasks);
  }

  replaceTask(id: number, task: Task) {
    const i = this.tasks.findIndex((task) => task.id === id);
    console.log('[DataApi] replaceTask', this.tasks[i], task);
    this.tasks[i] = task;
  }

  sortTasks() {
    this.tasks.sort(
      (taskA, taskB) => Number(taskB.lastTime) - Number(taskA.lastTime)
    );
  }

  /*
    When updating task times, we want to get the data, update the task object, then send
    but we don't have the full list anymore as we just sent it.
  */
  async fetchTaskTimes(taskId: number) {
    console.log('[DataApi] fetchTaskTimes');
    const task = this.tasks.find((task) => task.id == taskId);
    if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
    const taskTimes = await fetchTaskTimesRequest(taskId);
    task.taskTimes = taskTimes;
    this.tasks = [...this.tasks];
    this.updateTaskData(this.tasks);
  }
}

/*
  We need a return object with status, error, message, data
  on init we begin sending tasks data to the UI and send a loading state to the UI

  Loading (status) a temp object
  data
  success message | error message - closable messages, and a message history

  On the UI we want to see the most recent messages and be able to see a message history
  So somewhere we have a list of message, that all have a read state

  We can have a separate message api to handle the messages, which is data coming in from the top level,
  and displayed somewhere in the UI. Let's start by passing everything down, then change is to a context.
*/
