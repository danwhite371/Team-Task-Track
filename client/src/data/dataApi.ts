import type { Task } from '../types';
import {
  createTask,
  fetchTasks,
  startTask,
  stopTask,
  fetchTaskTimes,
} from './graphql';

export default class DataApi {
  updateTaskData: (tasks: Task[]) => void;
  tasks: Task[] = [];
  constructor(updateTaskData: (tasks: Task[]) => void) {
    this.updateTaskData = updateTaskData;
    this.sendTasks();
  }
  async createNewTask(name: string) {
    await createTask(name);
    await this.sendTasks();
  }
  async sendTasks() {
    const tasks = await fetchTasks();
    this.tasks = tasks;
    this.updateTaskData(tasks);
  }
  async startTask(id: number) {
    await startTask(id);
    await this.sendTasks();
  }

  async stopTask(id: number) {
    await stopTask(id);
    await this.sendTasks();
  }

  /*
    When updating task times, we want to get the data, update the task object, then send
    but we don't have the full list anymore as we just sent it.
  */
  async fetchTaskTimes(taskId: number) {
    const task = this.tasks.find((task) => task.id == taskId);
    if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
    const taskTimes = await fetchTaskTimes(taskId);
    task.taskTimes = taskTimes;
    this.tasks = [...this.tasks];
    this.updateTaskData(this.tasks);
  }
}

/*
  We need a return object with status, error, message, data
  on init we begin sending tasks data to the UI and send a loading state to the UI

  Loading (status) a temp object
  data
  success message | error message - closable messages, and a message history

  On the UI we want to see the most recent messages and be able to see a message history
  So somewhere we have a list of message, that all have a read state

  We can have a separate message api to handle the messages, which is data coming in from the top level,
  and displayed somewhere in the UI. Let's start by passing everything down, then change is to a context.
*/
