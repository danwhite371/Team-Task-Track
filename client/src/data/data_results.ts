import type { OperationResult } from '@/types';

const LoadingLoading: OperationResult = {
  status: 'loading',
  message: 'Loading...',
};
const taskLoadingSuccess: OperationResult = {
  status: 'success',
  message: 'Tasks loaded.',
};

const loadingConnection: OperationResult = {
  status: 'loading',
  message: 'Connecting...',
};
const taskCreatedSuccess: OperationResult = {
  status: 'success',
  message: 'Task created.',
};
const taskCreatedError: OperationResult = {
  status: 'error',
  message: 'Task name cannot be empty.',
};

export {
  LoadingLoading,
  taskLoadingSuccess,
  loadingConnection,
  taskCreatedSuccess,
  taskCreatedError,
};
