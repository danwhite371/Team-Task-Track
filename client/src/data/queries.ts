const taskQL = `id
      name
      active
      lastTime
      duration {
        milliseconds
        seconds
        minutes
        hours
        days
        years
      }
      secondsDuration
      createdAt
      updatedAt`;

const createTaskQuery = `
  mutation CreateTask($name: String!) {
    createTask(name: $name) {
      ${taskQL}
    }
  }
`;

const getTaskQuery = `
  query GetTask($id: Int!) {
    getTask(id: $id) {
      ${taskQL}
    }
  }
`;

const getAllTasksQuery = `query GetAllTasks {
  getAllTasks {
    ${taskQL}
  }
}`;

const startTaskQuery = `
  mutation StartTask($id: Int!) {
    startTask(id: $id) {
      ${taskQL}
    }
  }`;

const stopTaskQuery = `
  mutation StopTask($id: Int!) {
    stopTask(id: $id) {
      ${taskQL}
    }
  }`;

const deleteTaskQuery = `
  mutation DeleteTask($id: Int!) {
    deleteTask(id: $id) 
  }`;

const getTaskTimesQuery = `query GetTaskTimes($taskId: Int!){
    getTaskTimes(taskId: $taskId) {
      id
      start
      stop
      secondsDuration
    }
  }`;

const changeTaskNameQuery = `mutation ChangeTaskName($id: Int!, $name: String!) {
  changeTaskName(id: $id, name: $name) {
   name
    id
    id
    name
    active
    lastTime
    duration {
      milliseconds
      seconds
      minutes
      hours
      days
      years
    }
    secondsDuration
    createdAt
    updatedAt
  }
}`;

const queries = {
  createTaskQuery,
  getTaskQuery,
  getAllTasksQuery,
  startTaskQuery,
  stopTaskQuery,
  getTaskTimesQuery,
  deleteTaskQuery,
  changeTaskNameQuery,
};

export default queries;
