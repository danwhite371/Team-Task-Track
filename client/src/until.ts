import type { Duration, DurationValueType } from './types';

function formatDatetime(value: string | undefined): string | undefined {
  if (value == undefined) return;
  const date = new Date(Number(value));
  return date.toLocaleString();
}

const nullToZero = (value: number | undefined | null) =>
  value == null ? 0 : value;

const timeDuration = (diffInMs: number) => {
  const diffInSeconds = Math.floor(diffInMs / 1000);
  const diffInMinutes = Math.floor(diffInSeconds / 60);
  const diffInHours = Math.floor(diffInMinutes / 60);
  const diffInDays = Math.floor(diffInHours / 24);
  const years = Math.floor(diffInDays / 365);
  const milliseconds = diffInMs % 1000;
  const seconds = diffInSeconds % 60;
  const minutes = diffInMinutes % 60;
  const hours = diffInHours % 24;
  const days = diffInDays % 365;

  const duration = {
    milliseconds,
    seconds,
    minutes,
    hours,
    days,
    years,
  };
  return duration;
};

const durationToSecondsDuration = (duration: Duration): number => {
  let result: number = 0;
  if (duration.milliseconds) {
    result = duration.milliseconds / 1000;
  }
  if (duration.seconds) {
    result += duration.seconds;
  }
  if (duration.minutes) {
    result += duration.minutes * 60;
  }
  if (duration.hours) {
    result += duration.hours * 1440;
  }
  if (duration.years) {
    result += duration.years * 525600;
  }
  return result;
};

const durationToValueTypes = (duration: Duration) => {
  const valueTypes: DurationValueType[] = [];
  if (duration.years) valueTypes.push({ value: duration.years, type: 'y' });
  if (duration.days) valueTypes.push({ value: duration.days, type: 'd' });
  if (duration.hours) valueTypes.push({ value: duration.hours, type: 'h' });
  if (duration.minutes) valueTypes.push({ value: duration.minutes, type: 'm' });
  if (duration.seconds) valueTypes.push({ value: duration.seconds, type: 's' });
  return valueTypes;
};

const durationToString = (duration: Duration) => {
  const valueTypes = durationToValueTypes(duration);
  const result = valueTypes.map((vt) => `${vt.value}${vt.type}`).join(' ');
  // const result = valueTypes.flatMap((valueType, index) => {
  //   const vtString = `${valueType.value}${valueType.type}`;
  //   if(index === 0) {
  //     return [vtString]
  //   }
  //   return [' ', vtString]
  // })
  return result;
};

function lowercaseFirstChar(str: string): string {
  if (!str) {
    throw new Error('Unexpected null or empty string:');
  }
  return str.charAt(0).toLowerCase() + str.slice(1);
}

export {
  formatDatetime,
  nullToZero,
  timeDuration,
  lowercaseFirstChar,
  durationToSecondsDuration,
  durationToValueTypes,
  durationToString,
};
