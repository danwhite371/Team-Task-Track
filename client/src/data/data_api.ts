import type { OperationResult, Task } from '../types';
import { loadingConnection, LoadingLoading } from './data_results';
import {
  createTask,
  fetchTasks,
  startTask as startTaskRequest,
  stopTask as stopTaskRequest,
  fetchTaskTimes as fetchTaskTimesRequest,
} from './graphql';

export default class DataApi {
  updateTaskData: (tasks: Task[]) => void;
  updateOperationResult: (operationResult: OperationResult) => void;
  tasks: Task[] = [];

  constructor(
    updateTaskData: (tasks: Task[]) => void,
    updateOperationResult: (operationResult: OperationResult) => void
  ) {
    this.updateTaskData = updateTaskData;
    this.updateOperationResult = updateOperationResult;
    this.sendTasks();
  }

  async createNewTask(name: string) {
    this.updateOperationResult(loadingConnection);
    try {
      const task = await createTask(name);
      console.log('createNewTask', JSON.stringify(task, null, 2));
      this.tasks.unshift(task);
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task created.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async sendTasks() {
    this.updateOperationResult(LoadingLoading);
    try {
      const tasks = await fetchTasks();
      this.tasks = tasks;
      this.updateTaskData(tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Tasks loaded.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  // Get a new Task item with, replace current item with it, sort items by last time
  async startTask(id: number) {
    console.log('[DataApi] startTask', id);
    this.updateOperationResult(loadingConnection);
    try {
      const task = await startTaskRequest(id);
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task timer starter.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  async stopTask(id: number) {
    console.log('[DataApi] stopTask', id);
    this.updateOperationResult({ status: 'loading', message: 'Connecting...' });
    try {
      const task = await stopTaskRequest(id);
      console.log('[DataApi] stopTask', JSON.stringify(task, null, 2));
      this.replaceTask(id, task);
      this.sortTasks();
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task timer stopped.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  replaceTask(id: number, task: Task) {
    const i = this.tasks.findIndex((task) => task.id === id);
    console.log('[DataApi] replaceTask', this.tasks[i], task);
    this.tasks[i] = task;
  }

  sortTasks() {
    this.tasks.sort(
      (taskA, taskB) => Number(taskB.lastTime) - Number(taskA.lastTime)
    );
  }

  async fetchTaskTimes(taskId: number) {
    console.log('[DataApi] fetchTaskTimes');
    this.updateOperationResult({ status: 'loading', message: 'Loading...' });
    try {
      const task = this.tasks.find((task) => task.id == taskId);
      if (!task) throw new Error('[sendTaskTimes] Task id mismatch');
      const taskTimes = await fetchTaskTimesRequest(taskId);
      task.taskTimes = taskTimes;
      this.tasks = [...this.tasks];
      this.updateTaskData(this.tasks);
      this.updateOperationResult({
        status: 'success',
        message: 'Task times loaded.',
      });
    } catch (error: unknown) {
      this.updateOperationResult({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }
}
