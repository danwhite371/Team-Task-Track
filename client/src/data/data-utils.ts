import type { DataUtils } from '@/types';
import queries from './queries';

const { createTaskQuery, getAllTasksQuery, getTaskTimesQuery, startTaskQuery, stopTaskQuery } = queries;

export const dataUtils: DataUtils = {
  results: {
    loadingLoading: {
      status: 'loading',
      message: 'Loading...',
    },
    taskLoadingSuccess: {
      status: 'success',
      message: 'Tasks loaded.',
    },
    taskTimesLoadingSuccess: {
      status: 'success',
      message: 'Task times loaded.',
    },
    loadingConnection: {
      status: 'loading',
      message: 'Connecting...',
    },
    taskCreatedSuccess: {
      status: 'success',
      message: 'Task created.',
    },
    startTaskSuccess: {
      status: 'success',
      message: 'Task timer starter.',
    },
    stopTaskSuccess: {
      status: 'success',
      message: 'Task timer stopped.',
    },
    taskCreatedError: {
      status: 'error',
      message: 'Task name cannot be empty.',
    },
    failedFetchError: {
      status: 'error',
      message: 'Failed to fetch',
    },
  },
  requests: {
    getAllTasks: {
      operationName: 'GetAllTasks',
      query: getAllTasksQuery,
    },
    createTask: (name) => ({
      operationName: 'CreateTask',
      query: createTaskQuery,
      variables: { name },
    }),
    getTaskTimes: (taskId) => ({
      operationName: 'GetTaskTimes',
      query: getTaskTimesQuery,
      variables: { taskId },
    }),
    startTask: (id) => ({
      operationName: 'StartTask',
      query: startTaskQuery,
      variables: { id },
    }),
    stopTask: (id) => ({
      operationName: 'StopTask',
      query: stopTaskQuery,
      variables: { id },
    }),
  },
};
