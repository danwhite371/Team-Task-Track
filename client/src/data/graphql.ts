import { lowercaseFirstChar } from '@/until';
import {
  createTaskQuery,
  getAllTasksQuery,
  startTaskQuery,
  stopTaskQuery,
  getTaskTimesQuery,
} from './queries';
const GRAPHQL_URL = 'http://localhost:4000/';

type graphqlFetchProps = {
  query: string;
  operationName: string;
  variables?: any;
};
async function graphqlFetch(props: graphqlFetchProps) {
  console.log(`graphqlFetch\n${JSON.stringify(props, null, 2)}`);
  let responseData;
  try {
    const response = await fetch(GRAPHQL_URL, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify(props),
    });
    console.log(`response\n${JSON.stringify(response, null, 2)}`);
    responseData = await response.json();
    if (responseData.errors) {
      const message = responseData.errors[0].message;
      throw new Error(message);
    }
  } catch (error) {
    throw error;
  }

  console.log(`responseData\n${JSON.stringify(responseData, null, 2)}`);
  return responseData.data[lowercaseFirstChar(props.operationName)];
}

async function fetchTasks() {
  return await graphqlFetch({
    query: getAllTasksQuery,
    operationName: 'GetAllTasks',
  });
}

async function fetchTaskTimes(taskId: number) {
  const variables = { taskId };
  return await graphqlFetch({
    operationName: 'GetTaskTimes',
    query: getTaskTimesQuery,
    variables,
  });
}

async function createTask(name: string) {
  const variables = { name };
  return await graphqlFetch({
    operationName: 'CreateTask',
    query: createTaskQuery,
    variables,
  });
}

async function startTask(id: number) {
  const variables = { id };
  return await graphqlFetch({
    operationName: 'StartTask',
    query: startTaskQuery,
    variables,
  });
}

async function stopTask(id: number) {
  const variables = { id };
  return await graphqlFetch({
    operationName: 'StopTask',
    query: stopTaskQuery,
    variables,
  });
}

export { fetchTasks, fetchTaskTimes, createTask, startTask, stopTask };
