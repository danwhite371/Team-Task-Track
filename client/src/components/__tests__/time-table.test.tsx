import {
  cleanup,
  render,
  screen,
  within,
  act,
  waitFor,
} from '@testing-library/react';
import allTasks from '@/__fixtures__/GetAllTasks.json';
import taskTimesData from '@/__fixtures__/GetTaskTimes-91.json';
import taskTimesData92 from '@/__fixtures__/GetTaskTimes-92.json';
import DataApi from '@/data/data-api';
import { dataUtils } from '@/data/data-utils';
import type { Task, TaskTime } from '@/types';
import TimeTable from '../time-table';
import { durationToString, formatDatetime, timeDuration } from '@/until';

const tasks = allTasks as Task[];
const taskTimes = taskTimesData as TaskTime[];
const taskTimes92 = taskTimesData92 as TaskTime[];

global.fetch = jest.fn();
const mockUpdateTaskData = jest.fn();
const mockUpdateOperationResult = jest.fn();
const { results } = dataUtils;

beforeEach(() => {
  jest.clearAllMocks();
});

afterEach(() => {
  cleanup();
});

async function renderTaskTimes(taskId: number, taskTimes: TaskTime[]) {
  const taskData: Task | undefined = tasks.find((task) => task.id === taskId);
  if (taskData == null) throw new Error('taskData is null.');

  jest.spyOn(global, 'fetch').mockResolvedValueOnce({
    ok: true,
    json: () => Promise.resolve({ data: { getAllTasks: [taskData] } }),
  } as Response);

  const dataApi = await DataApi.create(
    mockUpdateTaskData,
    mockUpdateOperationResult
  );

  expect(mockUpdateTaskData).toHaveBeenCalledWith(
    expect.objectContaining([taskData])
  );
  expect(mockUpdateOperationResult).toHaveBeenNthCalledWith(
    1,
    expect.objectContaining(results.loadingLoading)
  );
  expect(mockUpdateOperationResult).toHaveBeenNthCalledWith(
    2,
    expect.objectContaining(results.taskLoadingSuccess)
  );

  jest.spyOn(global, 'fetch').mockResolvedValueOnce({
    ok: true,
    json: () => Promise.resolve({ data: { getTaskTimes: taskTimes } }),
  } as Response);

  jest.clearAllMocks();

  const { rerender } = render(
    <div data-testid="test-div">
      <TimeTable task={taskData} dataApi={dataApi} />
    </div>
  );

  taskData.taskTimes = taskTimes;

  await waitFor(() => {
    expect(mockUpdateTaskData).toHaveBeenCalledWith(
      expect.objectContaining([taskData])
    );
  });

  expect(mockUpdateOperationResult).toHaveBeenNthCalledWith(
    1,
    expect.objectContaining(results.loadingLoading)
  );
  expect(mockUpdateOperationResult).toHaveBeenNthCalledWith(
    2,
    expect.objectContaining(results.taskTimesLoadingSuccess)
  );

  await act(async () => {
    rerender(
      <div data-testid="test-div">
        <TimeTable task={taskData} dataApi={dataApi} />
      </div>
    );
  });

  let container = screen.getByTestId('test-div');
  expect(container).toMatchSnapshot();
}

describe('TimeTable', () => {
  it('should Render a TimeTable with active == false', async () => {
    await renderTaskTimes(91, taskTimes);

    const timeRow0 = screen.getByTestId('time-91-0');
    const timeRow1 = screen.getByTestId('time-91-1');
    const cells0 = within(timeRow0).getAllByRole('cell');
    const cells1 = within(timeRow1).getAllByRole('cell');
    expect(cells0[0]).toHaveTextContent(formatDatetime(taskTimes[0].start)!);
    expect(cells0[1]).toHaveTextContent(formatDatetime(taskTimes[0].stop!)!);
    let duration = durationToString(
      timeDuration(taskTimes[0].secondsDuration! * 1000)
    );
    expect(cells0[2]).toHaveTextContent(duration);
    expect(cells1[0]).toHaveTextContent(formatDatetime(taskTimes[1].start)!);
    expect(cells1[1]).toHaveTextContent(formatDatetime(taskTimes[1].stop!)!);
    duration = durationToString(
      timeDuration(taskTimes[1].secondsDuration! * 1000)
    );
    expect(cells1[2]).toHaveTextContent(duration);
  });

  it('should have empty content for stop and duration in the last row of an active Task', async () => {
    await renderTaskTimes(92, taskTimes92);
    const taskTimes = taskTimes92;
    const timeRow0 = screen.getByTestId('time-92-0');
    const timeRow1 = screen.getByTestId('time-92-1');
    const cells0 = within(timeRow0).getAllByRole('cell');
    const cells1 = within(timeRow1).getAllByRole('cell');
    expect(cells0[0]).toHaveTextContent(formatDatetime(taskTimes[0].start)!);
    expect(cells0[1]).toHaveTextContent(formatDatetime(taskTimes[0].stop!)!);
    let duration = durationToString(
      timeDuration(taskTimes[0].secondsDuration! * 1000)
    );
    expect(cells0[2]).toHaveTextContent(duration);
    expect(cells1[0]).toHaveTextContent(formatDatetime(taskTimes[1].start)!);
    expect(cells1[1]).toBeEmptyDOMElement();
    expect(cells1[2]).toBeEmptyDOMElement();
  });
});
