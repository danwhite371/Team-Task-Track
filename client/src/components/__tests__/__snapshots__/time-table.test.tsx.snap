// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeTable should Render a TimeTable with active == false 1`] = `
<div
  data-testid="test-div"
>
  <div
    class="relative overflow-x-auto"
    data-slot="table-container"
  >
    <table
      class="caption-bottom text-sm"
      data-slot="table"
    >
      <thead
        class="[&_tr]:border-b"
        data-slot="table-header"
      >
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
        >
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Start
          </th>
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Stop
          </th>
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Duration
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
        data-slot="table-body"
      >
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
          data-testid="time-91-0"
        >
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/19/2025, 8:08:15 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/19/2025, 8:08:31 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-right"
            data-slot="table-cell"
          >
            16
            <sup
              class="font-bold"
            >
              s
            </sup>
          </td>
        </tr>
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
          data-testid="time-91-1"
        >
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/19/2025, 8:08:39 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/20/2025, 10:43:07 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-right"
            data-slot="table-cell"
          >
            1
            <sup
              class="font-bold"
            >
              d
            </sup>
             
            2
            <sup
              class="font-bold"
            >
              h
            </sup>
             
            34
            <sup
              class="font-bold"
            >
              m
            </sup>
             
            27
            <sup
              class="font-bold"
            >
              s
            </sup>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`TimeTable should have empty content for stop and duration in the last row of an active Task 1`] = `
<div
  data-testid="test-div"
>
  <div
    class="relative overflow-x-auto"
    data-slot="table-container"
  >
    <table
      class="caption-bottom text-sm"
      data-slot="table"
    >
      <thead
        class="[&_tr]:border-b"
        data-slot="table-header"
      >
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
        >
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Start
          </th>
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Stop
          </th>
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-head"
          >
            Duration
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
        data-slot="table-body"
      >
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
          data-testid="time-92-0"
        >
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/19/2025, 8:19:04 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/20/2025, 10:45:21 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-right"
            data-slot="table-cell"
          >
            1
            <sup
              class="font-bold"
            >
              d
            </sup>
             
            2
            <sup
              class="font-bold"
            >
              h
            </sup>
             
            26
            <sup
              class="font-bold"
            >
              m
            </sup>
             
            16
            <sup
              class="font-bold"
            >
              s
            </sup>
          </td>
        </tr>
        <tr
          class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
          data-slot="table-row"
          data-testid="time-92-1"
        >
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          >
            6/20/2025, 10:45:30 AM
          </td>
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          />
          <td
            class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-right"
            data-slot="table-cell"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
